modules/compute/gke-node-pool/reservation_definitions.tf

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+data "google_compute_reservation" "specific_reservations" {
+  for_each = (
+    local.input_specific_reservations_count == 0 ?
+    {} :
+    {
+      for pair in flatten([
+        for zone in var.zones : [
+          for reservation in var.reservation_affinity.specific_reservations : {
+            key : "${coalesce(reservation.project, var.project_id)}/${zone}/${reservation.name}"
```

Comment:
nit: It looks like we are checking that specific reservation could be present in any of the zones present in the blueprint, I am wondering if we additionally ask user for zone in the reservation input, that will make this logic simpler and also user might catch their own mistakes easily.

Reply:
> It looks like we are checking that specific reservation could be present in any of the zones present in the blueprint

Yes.

> I am wondering if we additionally ask user for zone in the reservation input, that will make this logic simpler and also user might catch their own mistakes easily.

Actually, it was a deliberate choice. 

The question is whether to make the new variable (`var.reservation_affinity.specific_reservations[].zone`) a required variable  or not. If it were optional, our fallback would still be var.zones and in turn, we would require nested loops.

I chose to not introduce the variable at all in favor of keeping blueprints simple (one less variable) and not asking for input unless absolutely necessary.

One might question why was `var.reservation_affinity.specific_reservations[].projec`t variable introduced then? The project input is different from the zone input in that for making shared reservations work, we have no way of knowing the project which created the reservation. Whereas for zone, its value will always be one of `var.zones`.

Reply:
Thanks for the explanation!